_posts/2021-08-02-Musing-on-Microservices.md

```diff
@@ -0,0 +1,29 @@
+
+When it comes to availability, a similar school of thought applies. Unless different parts of your application have different availability requirements, can you just run multiple copies of your monolith? And even if you have parts of your application that have high availability requirements will moving to microservices increase your availability? Running microservices in production can be difficult and introduces several more infrastructure components to worry about (e.g. a Kubernetes backplane) which, if they go down could reduce your availability. It is worth considering whether or not you really need such high availability and whether it is worth the cost of trying to achieve it? Perhaps it costs your business less to go down every now and again than it does to hire and maintain a team of DevOps engineers.
+
+It's also important to think about compound failure when dealing with microservices. With a monolith, if at least one copy of your monolith is up 99.9% of the time (ignoring external dependencies such as databases) then your uptime is 99.9%. Now, imagine you have a microservices architecture with two services A and B which must both be up for the user to do anything. If both of these services are up 99.9% of the time then your total uptime is actually 0.999 x 0.999 x 100 = 99.8%. This difference may seem small but given that these compounds for each dependency you have, you can quickly have much lower availability than you thought. 
+
+The other thing to consider when thinking about availability is failure modes. In a microservices architecture, you have more individual components, which means you have more combinations of things that can go wrong. This often makes it harder to reason about behaviour in the case of a bug or a crash. For example, if one service (service A) goes down but another (service B) stays up, on one hand, you can keep serving requests to service B, however, service A and B may now be out of sync with each other (for example, service B has consumed some event that service A hasn't) which now may lead to unexpected and confusing behaviour. In many cases, it is actually better for a failure to mean that your application just stops, rather than half processing requests.
+
+The final thing to think about with microservices is what I'm calling: microservices as a design pattern. In some companies, microservices have become a way of managing code complexity. In some codebases where you might have a class, in others, you may now have a microservice. Now while this approach can work, if you are considering adopting microservices, for this reason, I would strongly suggest you consider whether or not it is worth the overhead of having to support such a high number of microservices.
```

Comment:
Could you talk about alternatives here?

e.g. more effective abstraction (interfaces? better docs?), separating out code ownership by package?